oprf/oprf.go

```diff
@@ -0,0 +1,267 @@
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   string
+	K     *KeyPair
+}
+
+// ServerContext implements the functionality of a Server.
+// TODO: add FullEvaluate
+type ServerContext interface {
+	// Evaluate evaluates the token
+	Evaluate(b BlindToken) (*Evaluation, error)
+}
+
+func generateCtx(suiteID uint16) string {
+	ctx := "000"
+	ctx = ctx + fmt.Sprintf("%x", suiteID)
+
+	return ctx
+}
+
+func generateKeys(suite *group.Ciphersuite) (*KeyPair, error) {
+	privK, err := suite.RandomScalar()
+	if err != nil {
+		return nil, err
+	}
+
+	pubK, err := suite.ScalarMultBase(privK)
+	if err != nil {
+		return nil, err
+	}
+
+	return &KeyPair{pubK, privK}, nil
+}
+
+func suiteFromID(suiteID uint16) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	// TODO: add other suites
+	switch suiteID {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256")
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, nil
+}
+
+// NewServer creates a new instantiation of a Server.
+func NewServer(suiteID uint16) (*Server, error) {
+	suite, err := suiteFromID(suiteID)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := generateCtx(suiteID)
+
+	keyPair, err := generateKeys(suite)
+	if err != nil {
+		return nil, err
+	}
+
+	server := &Server{}
+	server.suite = suite
+	server.ctx = ctx
+	server.K = keyPair
+
+	return server, nil
+}
+
+// Evaluate creates an evaluation of the blided token.
+func (s *Server) Evaluate(b BlindToken) (*Evaluation, error) {
+	p := group.NewPoint(s.suite)
+	p.Deserialize(b)
+
+	z, err := p.ScalarMult(s.K.PrivK)
+	if err != nil {
+		return nil, err
+	}
+
+	ser, err := z.Serialize()
+	if err != nil {
+		return nil, err
+	}
+
+	eval := &Evaluation{ser}
+	return eval, nil
+}
+
+// NewClient creates a new instantiation of a Client.
+func NewClient(suiteID uint16) (*Client, error) {
+	suite, err := suiteFromID(suiteID)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := generateCtx(suiteID)
+
+	client := &Client{}
+	client.suite = suite
+	client.ctx = ctx
+
+	return client, nil
+}
+
+// Blind generates a token and blinded data.
+func (c *Client) Blind(in []byte) (*Token, BlindToken, error) {
+	// sample a random blind
+	r, err := c.suite.RandomScalar()
+	if err != nil {
+		return nil, nil, errors.New("failed at blinding")
+	}
+
+	p, err := c.suite.HashToGroup(in)
+	if p == nil {
+		return nil, nil, errors.New("failed at blinding")
+	}
+
+	t, err := p.ScalarMult(r)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bToken, err := t.Serialize()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	token := &Token{in, r}
+	return token, bToken, nil
+}
+
+// Unblind unblinds the server response.
+func (c *Client) Unblind(t *Token, e *Evaluation) (IssuedToken, error) {
+	p := group.NewPoint(c.suite)
+	p.Deserialize(e.element)
+
+	r := t.blind
+	n := c.suite.Order()
+	// TODO: as to not export the internal field of the scalar, this can be encapsulated
```

Comment:
👍 Yeah, great idea! Why not add a `ModInverse` function to the `Scalar` interface?

Reply:
Yes! Just did it.